fix(sqllab): avoid unexpected re-rendering on DatabaseSelector

justinpark opened the pull request:
### SUMMARY

Reopen: #21141

Since DatabaseSelector component has an unnecessary key, DatabaseSelector will be re-mounted  rather than update whenever currentDatabase object is updated. This causes the unexpected duplicate database list api request.
This commit removes this key so will skip unnecessary re-rendering and data fetching.
It also fixes the logic for currentDb that directly transform from db instead of useState. (This resolves the issue for #21174)

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Before:

![fix--unnecessary-dabaseselector-rerendering](https://user-images.githubusercontent.com/1392866/185713445-a22e86e1-9382-4ecc-81a4-bdb692d016d1.gif)

After:

![Screen Shot 2022-08-19 at 2 50 12 PM](https://user-images.githubusercontent.com/1392866/185713454-b6a05d82-2162-4b6d-9fab-13d36a209401.png)

### TESTING INSTRUCTIONS
go to Sql Editor and check the network list

### ADDITIONAL INFORMATION
- [x] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

cc: @ktmud

justinpark committed:
fix(sqllab): avoid unexpected re-rendering on DatabaseSelector

michael-s-molina commented on superset-frontend/src/components/Select/AsyncSelect.tsx:
```diff
@@ -690,7 +690,7 @@ const AsyncSelect = forwardRef(
         }
       }
-    }, [loadingEnabled, fetchPage, allowFetch, inputValue, debouncedFetchPage]);
+    }, [loadingEnabled, allowFetch, inputValue]);
```
```suggestion
    }, [loadingEnabled, fetchPage, allowFetch, inputValue, debouncedFetchPage]);
```

michael-s-molina reviewed (changes requested):
Thanks for the PR @justinpark. Is there a reason for removing `AsyncSelect` `useEffect` dependencies?

justinpark commented:
> Is there a reason for removing AsyncSelect useEffect dependencies?

@michael-s-molina Because fetchPage kept changing while initializing due to the object dependencies, useEffect has been triggered multiple times which causes unnecessary `fetchPage` calls.

I made [a change](https://github.com/apache/superset/pull/21316/commits/243ba07e55d24eb3c007f58557ddde68dcd139e9) to reduce and fix the dependencies for fetchPage and rollback this useEffect dependencies.

michael-s-molina commented:
> > Is there a reason for removing AsyncSelect useEffect dependencies?
> 
> @michael-s-molina Because fetchPage kept changing while initializing due to the object dependencies, useEffect has been triggered multiple times which causes unnecessary `fetchPage` calls.
> 
> I made [a change](https://github.com/apache/superset/pull/21316/commits/243ba07e55d24eb3c007f58557ddde68dcd139e9) to reduce and fix the dependencies for fetchPage and rollback this useEffect dependencies.

@justinpark We can't add the responsibility of the immutability of the `options` to the Select component because there are some cases where we want to fire the requests again when the function changes. To fix your problem, you just need to make sure `options` only change when you in fact want them to. This is generally accomplished by the use of `useMemo` or `useCallback` on the parent component depending on the case.

justinpark commented:
> > > Is there a reason for removing AsyncSelect useEffect dependencies?
> > 
> > 
> > michael-s-molina Because fetchPage kept changing while initializing due to the object dependencies, useEffect has been triggered multiple times which causes unnecessary `fetchPage` calls.
> > I made [a change](https://github.com/apache/superset/pull/21316/commits/243ba07e55d24eb3c007f58557ddde68dcd139e9) to reduce and fix the dependencies for fetchPage and rollback this useEffect dependencies.
> 
> justinpark We can't add the responsibility of the immutability of the `options` to the Select component because there are some cases where we want to fire the requests again when the function changes. To fix your problem, you just need to make sure `options` only change when you in fact want them to. This is generally accomplished by the use of `useMemo` or `useCallback` on the parent component depending on the case.

Okay. @michael-s-molina   I finally cleaned up the main cause.

1. I removed [the unnecessary sort function dependency](https://github.com/apache/superset/blob/master/superset-frontend/src/components/Select/AsyncSelect.tsx#L319) on mergeData because fetchOptions resolver only cares the length(so sorting is unnecessary and heavy). 

https://github.com/apache/superset/blob/b739e27f6dc4b159d766074e3e353a5546d00adb/superset-frontend/src/components/Select/AsyncSelect.tsx#L340-L346

2. I fixed [the invalid checking](https://github.com/apache/superset/blob/master/superset-frontend/src/components/Select/AsyncSelect.tsx#L345) for setAllValuesLoaded. (Since allValuesLoaded should check whether the current request is for the full list or not, it should check the `search` rather than `value)

3. When removes the key in DatabaseSelector, its initial database selection has gone. (#21174) The issue is caused by the missing currentDb mapping when db prop is updated. I added the useEffect logic to handle this issue.

https://github.com/apache/superset/blob/b739e27f6dc4b159d766074e3e353a5546d00adb/superset-frontend/src/components/DatabaseSelector/index.tsx#L135-L144 

4. Lastly I found the same useEffect logic on AsyncSelect so removed one

https://github.com/apache/superset/blob/b739e27f6dc4b159d766074e3e353a5546d00adb/superset-frontend/src/components/Select/AsyncSelect.tsx#L473-L477

https://github.com/apache/superset/blob/b739e27f6dc4b159d766074e3e353a5546d00adb/superset-frontend/src/components/Select/AsyncSelect.tsx#L513-L517

justinpark committed:
revert back sortComparatorForNoSearch and use selectValueRef

michael-s-molina reviewed (approved):
LGTM. Thanks for the fixes @justinpark!
